File: Sources/AINotebookCore/FollowupSuggester.swift
import Foundation

/// Asks the chat model for up to 3 short follow-up questions a user might ask
/// next, given their question and the assistant's answer. Dependency-light:
/// reuses the same `ChatStreaming` abstraction the chat engine uses.
public actor FollowupSuggester {
    private let chat: ChatStreaming
    public let chatModel: String

    public init(chat: ChatStreaming, chatModel: String) {
        self.chat = chat
        self.chatModel = chatModel
    }

    /// Up to 3 trimmed, non-empty follow-up question strings, parsed from the
    /// model output (one per line, leading list markers stripped).
    public func generate(userText: String, answer: String) async throws -> [String] {
        let prompt = """
        Based on this question and answer, suggest 3 short, specific follow-up \
        questions the user might ask next. One question per line, no numbering.

        Question:
        \(userText)

        Answer:
        \(answer)
        """
        let turns: [ChatTurn] = [ChatTurn(role: .user, content: prompt)]
        var assembled = ""
        for try await token in chat.stream(model: chatModel, messages: turns) {
            assembled += token
        }
        return Self.parse(assembled)
    }

    /// Split into lines, strip leading list markers ("1.", "-", "•"), drop
    /// blanks, cap at 3.
    static func parse(_ raw: String) -> [String] {
        var out: [String] = []
        for line in raw.split(separator: "\n", omittingEmptySubsequences: false) {
            let stripped = stripMarker(String(line))
                .trimmingCharacters(in: .whitespacesAndNewlines)
            guard !stripped.isEmpty else { continue }
            out.append(stripped)
            if out.count == 3 { break }
        }
        return out
    }

    /// Remove a leading ordered/unordered list marker like "1.", "2)", "-",
    /// "*", or "•" from the start of a line.
    private static func stripMarker(_ line: String) -> String {
        var s = Substring(line.trimmingCharacters(in: .whitespaces))
        if let first = s.first, first == "-" || first == "*" || first == "•" {
            s = s.dropFirst()
            return String(s)
        }
        // Numbered markers: leading digits followed by "." or ")".
        var idx = s.startIndex
        while idx < s.endIndex, s[idx].isNumber { idx = s.index(after: idx) }
        if idx > s.startIndex, idx < s.endIndex, s[idx] == "." || s[idx] == ")" {
            return String(s[s.index(after: idx)...])
        }
        return String(s)
    }
}

File: windows/src/AINotebook.Core/Rag/FollowupSuggester.cs
using AINotebook.Core.Models;
using AINotebook.Core.Ollama;

namespace AINotebook.Core.Rag;

/// <summary>
/// Asks the chat model for up to 3 short follow-up questions a user might ask
/// next, given their question and the assistant's answer. Dependency-light:
/// reuses the same <see cref="IChatStreaming"/> abstraction the chat engine uses.
/// </summary>
public sealed class FollowupSuggester
{
    private readonly IChatStreaming _chat;
    public string ChatModel { get; }

    public FollowupSuggester(IChatStreaming chat, string chatModel)
    {
        _chat = chat;
        ChatModel = chatModel;
    }

    /// <summary>
    /// Up to 3 trimmed, non-empty follow-up question strings, parsed from the
    /// model output (one per line, leading list markers stripped).
    /// </summary>
    public async Task<IReadOnlyList<string>> GenerateAsync(
        string userText, string answer, CancellationToken ct = default)
    {
        var prompt =
            "Based on this question and answer, suggest 3 short, specific follow-up\n" +
            "questions the user might ask next. One question per line, no numbering.\n\n" +
            "Question:\n" + userText + "\n\n" +
            "Answer:\n" + answer;

        var turns = new List<ChatTurn> { new(ChatRole.User, prompt) }; // NO system turn
        var assembled = "";
        await foreach (var token in _chat.StreamAsync(ChatModel, turns, ct))
            assembled += token;

        return Parse(assembled);
    }

    /// <summary>
    /// Split into lines, strip leading list markers ("1.", "-", "•"), drop
    /// blanks, cap at 3.
    /// </summary>
    internal static IReadOnlyList<string> Parse(string raw)
    {
        var output = new List<string>();
        foreach (var line in raw.Split('\n'))
        {
            var stripped = StripMarker(line).Trim();
            if (stripped.Length == 0) continue;
            output.Add(stripped);
            if (output.Count == 3) break;
        }
        return output;
    }

    /// <summary>
    /// Remove a leading ordered/unordered list marker like "1.", "2)", "-",
    /// "*", or "•" from the start of a line.
    /// </summary>
    private static string StripMarker(string line)
    {
        var s = line.Trim();
        if (s.Length == 0) return s;
        var first = s[0];
        if (first == '-' || first == '*' || first == '•')
            return s[1..];
        // Numbered markers: leading digits followed by "." or ")".
        var idx = 0;
        while (idx < s.Length && char.IsDigit(s[idx])) idx++;
        if (idx > 0 && idx < s.Length && (s[idx] == '.' || s[idx] == ')'))
            return s[(idx + 1)..];
        return s;
    }
}
